dask_geopandas/core.py

```diff
@@ -429,6 +432,100 @@ def geohash(self, string=True, p=12):
 
         return geohashes
 
+    def spatial_shuffle(
+        self,
+        by="hilbert",
+        p=None,
+        calculate_partitions=True,
+        npartitions=None,
+        divisions=None,
+        **kwargs,
+    ):
+        """
+        Shuffle the data into spatially consistent partitions.
+
+        This realigns the dataset to be spatially sorted, i.e. geometries that are
+        spatially near each other together will be within the same partition. This is
+        useful especially for overlay operations like a spatial join as it reduces the
+        number of interactions between individual partitions.
+
+        The spatial information is stored in the index and will replace the existing
+        index.
+
+        Note that ``spatial_shuffle`` uses ``set_index`` under the hood and comes with
+        all its potential performance drawbacks.
+
+        Parameters
+        ----------
+        by : string (default 'hilbert')
+            sorting method, one of {'hilbert', 'morton, 'geohash'}. See
+            ``hilbert_distance``, ``morton_distance`` and ``geohash`` methods for
+            details.
+        p : int (default None)
+            precision of the sorting method. Defaults to 15 for Hilbert and Morton
+            curves and 12 for geohash.
+        calculate_partitions : bool (default True)
+            calculate new spatial partitions after shuffling
+        npartitions : int, None, or 'auto'
+            The ideal number of output partitions. If None, use the same as the input.
+            If 'auto' then decide by memory use. Only used when divisions is not given.
+            If divisions is given, the number of output partitions will be
+            len(divisions) - 1.
+        divisions: list, optional
+            The “dividing lines” used to split the new index into partitions. Needs to
+            match the values returned by the sorting method.
+
+        **kwargs
+            keyword arguments passed to ``set_index``.
+
+        Returns
+        -------
+        dask_geopandas.GeoDataFrame
+
+        Notes
+        -----
+        This method, similarly to ``calculate_spatial_partitions``, is computed
+        partially eagerly as it needs to load the active geometry column and compute
+        spatial partitions.
```

Comment:
This is the case when `calculate_spatial_partitions=True`? If so, maybe mention in the above paragraph you can make it fully lazily by passing that keyword.  

And for example, I suppose that if you do a directly following write to disk, such as `ddf.spatial_shuffle().to_parquet()`, you actually want to turn this off? Because then dask can combine all things in a single graph, and we will calculate the bounds of each partition anyway when writing to Parquet.

Reply:
> This is the case when calculate_spatial_partitions=True? If so, maybe mention in the above paragraph you can make it fully lazily by passing that keyword.

I don't think so, it seems to be computed eagerly no matter the `calculate_spatial_partitions` option because dask needs to know the divisions for the new index. And to get those, it needs to load geometry and compute Hilbert curve.

Reply:
Yeah, the `set_index` documentation (https://docs.dask.org/en/latest/generated/dask.dataframe.DataFrame.set_index.html) has this about the `compute` keyword:

> Whether or not to trigger an immediate computation. Defaults to False. Note, that even if you set `compute=False`, an immediate computation will still be triggered if `divisions` is `None`.

Reply:
Maybe we can then clarify the note a little bit? Eg "... as it needs to calculate the distances for all existing partitions before it can determine the divisions for the new spatially-shuffled partitions"